src/object/cli_obj.c

```diff
@@ -2149,10 +2158,28 @@ shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
 static int
 shard_io_task(tse_task_t *task)
 {
-	struct shard_auxi_args		*shard_auxi;
+	struct shard_auxi_args	*shard_auxi;
+	daos_handle_t		 th;
+	int			 rc;
 
 	shard_auxi = tse_task_buf_embedded(task, sizeof(*shard_auxi));
 
+	/*
+	 * If this task belongs to a TX, and if the epoch we got earlier
+	 * doesn't contain a "chosen" TX epoch, then we may need to reinit the
+	 * task. (See dc_tx_get_epoch.) Because tse_task_reinit is less
+	 * practical in the middle of a task, we do it here at the beginning of
+	 * shard_io_task.
+	 */
+	th = shard_auxi->obj_auxi->th;
+	if (daos_handle_is_valid(th) && !dtx_epoch_chosen(&shard_auxi->epoch)) {
```

Comment:
Only for fetch ? For update and punch cases, we just cache on the client without consider the epoch before TX commit.

Reply:
I needed to remove the "is-a-read-operation" condition for non_cpd cases in EP1. I think the removal is harmless because for cpd cases update/punch operations should not reach this code path, while CPD (if it goes through here) should get choosing and later set epoch for daos_tx_hdl2epoch calls after commit.